compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -0,0 +1,306 @@
+use crate::{ImplTraitPosition, ResolverAstLoweringExt};
+
+use super::{ImplTraitContext, LoweringContext, ParamMode};
+
+use ast::visit::Visitor;
+use hir::def::{DefKind, PartialRes, Res};
+use hir::{BodyId, HirId};
+use rustc_ast as ast;
+use rustc_ast::*;
+use rustc_hir as hir;
+use rustc_hir::def_id::DefId;
+use rustc_middle::span_bug;
+use rustc_middle::ty::ResolverAstLowering;
+use rustc_span::symbol::kw;
+use rustc_span::{symbol::Ident, Span};
+use rustc_target::spec::abi;
+use std::iter;
+
+pub struct DelegationResults<'hir> {
+    pub body_id: hir::BodyId,
+    pub sig: hir::FnSig<'hir>,
+    pub generics: &'hir hir::Generics<'hir>,
+}
+
+impl<'hir> LoweringContext<'_, 'hir> {
+    pub fn delegation_has_self(&self, node_id: NodeId, span: Span) -> bool {
+        let res_id = self.get_delegation_res_id(node_id, span);
+        let Some(res_id) = res_id else {
+            return false;
+        };
+
+        if let Some(local_res_id) = res_id.as_local() {
+            self.resolver.has_self.contains(&local_res_id)
+        } else {
+            match self.tcx.def_kind(res_id) {
+                DefKind::Fn => false,
+                DefKind::AssocFn => self.tcx.associated_item(res_id).fn_has_self_parameter,
+                _ => span_bug!(span, "unexpected DefKind for delegation item"),
+            }
+        }
+    }
+
+    pub fn lower_delegation(
+        &mut self,
+        delegation: &Delegation,
+        node_id: NodeId,
```

Comment:
It is not clear what these `node_id`s passed around mean, the naming is very generic.
It is also not clear where the selection between the `resolution_of(item_id)` for trait impl items, and `resolution_of(delegation.id)` for all other items happens.
Please make it clear and commented.
(The selection should fully happen here in AST lowering, and not in resolve.)

Reply:
I suggest using some consistent naming scheme to improve this, e.g.
- `item_id` for ids of delegation items
- `path_id` for ids of delegation paths (may be unnecessary because the is always `delegation.id`)
- `sig_id` for ids of items from which the signature is inherited